std/experimental/typecons.d

```diff
@@ -1074,3 +1074,307 @@ pure nothrow @system unittest
+        _ok = false;
+    }
+    /// ditto
+    this(ref Unexpected!E unexpectedValue)
+    {
+        _store.unexpectedValue = unexpectedValue; // TODO do we need to care about `_store.unexpectedValue` init value??
+        _ok = false;
+    }
+
+    /// Assign from expected value `expectedValue.`
+    void opAssign()(auto ref T expectedValue) @trusted
+    {
+        // TODO is this ok?:
+        clear();
+        moveEmplace(expectedValue, _store.expectedValue);
+        _ok = true;
+    }
+
+    /// Assign from unexpected value `unexpectedValue.`
+    void opAssign(E unexpectedValue) @trusted
+    {
+        clear();
+        moveEmplace(unexpectedValue, _store.unexpectedValue);
+        _ok = false;
+    }
+
+    /// Clear (empty) contents.
+    private void clear() @trusted
+    {
+        release();
+        static if (isAddress!T)
+        {
+            _store.expectedValue = null;
+        }
+    }
+
+    /// Release any memory used to store contents.
+    private void release() @trusted
+    {
+        import std.traits : hasElaborateDestructor;
+        if (hasExpectedValue)
+        {
+            static if (!is(T == class))
+            {
+                static if (hasElaborateDestructor!T)
+                {
+                    destroy(_store.expectedValue);
+                }
+            }
+            _ok = false;
+        }
+        else
+        {
+            static if (!is(E == class))
+            {
+                static if (hasElaborateDestructor!E)
+                {
+                    destroy(_store.unexpectedValue);
+                }
+            }
+            destroy(_store.unexpectedValue);
+            // TODO change _ok?
+        }
+    }
+
+    /** Is `true` iff this has a expectedValue of type `T`. */
+    bool hasExpectedValue() const
+    {
+        return _ok;
+    }
+
+    import std.functional : unaryFun;
+
+    /** If `this` is an expected value (of type `T`) apply `fun` on it and
+     * return result, otherwise return current unexpected value (of type `E`).
+     *
+     * See_Also: https://dlang.org/phobos/std_typecons.html#.apply
+     */
+    Expected!(typeof(unaryFun!fun(T.init)), E) apply(alias fun)() @trusted
+    {
+        alias fn = unaryFun!fun;
+        if (hasExpectedValue)
+        {
+            return typeof(return)(fn(_store.expectedValue));
+        }
+        else
+        {
+            return typeof(return)(Unexpected!E(_store.unexpectedValue));
+        }
+    }
+
+    // range interface:
+
+    /// Check if empty.
+    @property bool empty() const
+    {
+        return !_ok;
+    }
+
+    /// Get current value.
+    @property inout(T) front() inout @trusted
+    {
+        assert(_ok);
+        return _store.expectedValue;
+    }
+
+    /// Pop (clear) current value.
+    void popFront()
+    {
+        assert(_ok);
+        clear();
+    }
+
+private:
+    union Store // TODO can we be sure this union is zero-initialized before constructor call?
```

Comment:
This question is answered in the D language specification: https://dlang.org/spec/struct.html#default_union_init